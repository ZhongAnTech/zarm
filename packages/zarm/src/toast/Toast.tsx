import React, { useCallback, useEffect, useRef, useState } from 'react';
import BaseToastProps from './interface';
import Popup from '../popup';
import { ConfigContext } from '../n-config-provider';

export interface ToastProps extends BaseToastProps {
  className?: string;
}
export interface UseToast {
  show: (props: ToastProps) => void;
  hide: () => void;
}
interface CompoundedComponent
  extends React.ForwardRefExoticComponent<ToastProps & React.RefAttributes<HTMLElement>> {
  useToast: () => UseToast;
}
const Toast = React.forwardRef<unknown, ToastProps>((props, ref) => {
  const { prefixCls: globalPrefixCls } = React.useContext(ConfigContext);
  const prefixCls = `${globalPrefixCls}-toast`;

  const toastRef = (ref as any) || React.createRef<HTMLElement>();
  const { className, stayTime, content, visible: propVisible, afterClose, ...others } = props;

  const [visible, setVisible] = useState(propVisible!);

  const timerRef = useRef<NodeJS.Timeout>();

  const _hide = () => {
    setVisible(false);
  };

  const autoClose = useCallback(() => {
    if (stayTime! > 0) {
      timerRef.current = setTimeout(() => {
        _hide();
        clearTimeout(timerRef.current!);
      }, stayTime);
    }
  }, [stayTime]);

  useEffect(() => {
    if (visible) {
      autoClose();
    }
    return () => {
      clearTimeout(timerRef.current!);
    };
  }, [autoClose, visible]);

  useEffect(() => {
    setVisible(propVisible!);
  }, [propVisible]);

  return (
    <Popup
      direction="center"
      maskType="transparent"
      width="70%"
      {...others}
      visible={visible}
      afterClose={afterClose}
    >
      <div className={prefixCls} ref={toastRef}>
        <div className={`${prefixCls}__container`}>{content}</div>
      </div>
    </Popup>
  );
}) as CompoundedComponent;

Toast.displayName = 'Toast';

Toast.defaultProps = {
  visible: false,
  stayTime: 3000,
  mask: false,
};

export default Toast;
