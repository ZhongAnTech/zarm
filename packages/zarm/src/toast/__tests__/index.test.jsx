import React from 'react';
import { mount } from 'enzyme';
import Toast from '../index';

describe('Toast', () => {
  it('renders correctly', () => {
    const wrapper = mount(<Toast visible>foo</Toast>);
    expect(wrapper).toMatchSnapshot();
  });

  it('visible change true', () => {
    const wrapper = mount(<Toast />);
    wrapper.setProps({ visible: true });
    expect(wrapper).toMatchSnapshot();
  });

  // it('visible change false', () => {
  //   const afterClose = jest.fn();
  //   const wrapper = mount(<Toast visible afterClose={afterClose} />);
  //   wrapper.setProps({ visible: false });
  //   wrapper.simulate('transitionEnd');
  //   wrapper.simulate('animationEnd');
  //   expect(wrapper.state('visible')).toEqual(false);
  //   expect(wrapper).toMatchSnapshot();
  // });

  // it('static function show', () => {
  //   jest.useFakeTimers();
  //   Toast.show('toast内容');
  //   jest.runAllTimers();
  //   const toastContainer = document.getElementsByClassName('za-toast-container');
  //   expect(toastContainer.length).toEqual(1);
  // });

  // it('static function pass params', () => {
  //   jest.useFakeTimers();
  //   Toast.show({
  //     content: 'toast内容',
  //   });
  //   jest.runAllTimers();
  //   const toastContainer = document.getElementsByClassName('za-toast-container');
  //   expect(toastContainer.length).toEqual(1);
  // });

  // it('static function hide', () => {
  //   jest.useFakeTimers();
  //   Toast.show();
  //   jest.runAllTimers();
  //   Toast.hide();
  //   // todo：实现静态方法的动画结束
  //   // const toastContainer = document.getElementsByClassName('toast-container');
  //   // do sth to travel transform
  //   // expect(toastContainer.length).toEqual(0);
  // });
});
