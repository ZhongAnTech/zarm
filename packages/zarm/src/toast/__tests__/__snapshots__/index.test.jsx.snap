// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast renders correctly 1`] = `ReactWrapper {}`;

exports[`Toast visible change true 1`] = `ReactWrapper {}`;
